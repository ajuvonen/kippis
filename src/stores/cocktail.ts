import {
  getByIngredients,
  getBySearchString,
  getDetails,
  getNonAlcoholic,
  getByFirstLetter,
} from '@/api';
import {SEARCHABLE_ALCOHOLS} from '@/utils/constants';
import type {FullDetailsCocktail, SearchResultCocktail, SearchStringProps} from '@/utils/types';
import {defineStore} from 'pinia';
import {compose, flatten, map, prop, sortBy, uniq, without} from 'ramda';

export const useCocktailStore = defineStore('cocktail', {
  state: () => ({
    currentSearch: {} as SearchStringProps,
    cocktailDetails: [] as FullDetailsCocktail[],
    preventFetch: false,
    searchResults: [] as SearchResultCocktail[],
    selection: new Set<number>(),
    highlightedCocktail: null as number | null,
  }),
  getters: {
    getCocktailDetails: (state) => (id: number) => state.cocktailDetails.find((cocktail) => cocktail.id === id) || null,
    // Get a "shopping list" of ingredients. They should be unique, sorted and without some obvious items
    getAllIngredients(state): string[] {
      return compose(
        sortBy(prop(0)),
        without(['water', 'ice']),
        uniq,
        flatten<string[][]>,
        map((id: number) =>
          // Make cream, egg, and citrus fruits show up as their base form 
          this.getCocktailDetails(id)!.ingredients.map(({ingredient}) =>
            ingredient.replace(/^whipped | peel$| spiral$| white$| yolk$/, ''),
          ),
        ),
      )(Array.from(state.selection));
    },
  },
  actions: {
    async addToSelection(id: number) {
      await this.fetchCocktail(id);
      this.selection.add(id);
    },
    async fetchCocktail(id: number) {
      if (!this.cocktailDetails.some((cocktail) => cocktail.id === id)) {
        this.cocktailDetails.push(await getDetails(id));
      }
    },
    async openCocktailModal(id: number) {
      await this.fetchCocktail(id);
      this.highlightedCocktail = id;
    },
    removeFromSelection(id: number) {
      this.selection.delete(id);
    },
    async search(searchString: string) {
      if (searchString.length === 1) {
        this.searchResults = await getByFirstLetter(searchString);
      } else {
        this.searchResults = await getBySearchString(searchString);
      }
    },
    async searchNonAlcoholic() {
      this.searchResults = await getNonAlcoholic();
    },
    async searchWithTag(tag: string) {
      this.searchResults = await getByIngredients(
        SEARCHABLE_ALCOHOLS.find((alcohol) => tag === alcohol.tag)?.ingredients!,
      );
    },
  },
});
