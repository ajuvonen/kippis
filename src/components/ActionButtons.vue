<script setup lang="ts">
import {storeToRefs} from 'pinia';
import {useI18n} from 'vue-i18n';
import {useCocktailStore} from '@/stores/cocktail';
import IconComponent from '@/components/IconComponent.vue';
import type {SearchResultCocktail} from '@/utils/types';

defineProps<{
  cocktail: SearchResultCocktail;
}>();

const {t} = useI18n();
const cocktailStore = useCocktailStore();
const {selection} = storeToRefs(cocktailStore);
const {addToSelection, removeFromSelection} = cocktailStore;
</script>
<template>
  <button
    v-if="!selection.has(cocktail.id)"
    :aria-label="t('searchResults.addCocktail', [cocktail.name])"
    :data-test-id="`add-cocktail-${cocktail.id}`"
    class="action-button"
    @click="addToSelection(cocktail.id)"
  >
    <IconComponent icon="plus" />
  </button>
  <button
    v-else
    :aria-label="t('searchResults.removeCocktail', [cocktail.name])"
    :data-test-id="`remove-cocktail-${cocktail.id}`"
    class="action-button bg-rose-400 border-slate-800"
    @click="removeFromSelection(cocktail.id)"
  >
    <IconComponent icon="trashCan" />
  </button>
</template>
<style lang="scss" scoped>
.action-button {
  @apply absolute bottom-2 right-2 p-0 m-0 rounded-md w-10 text-lg;
}
</style>
