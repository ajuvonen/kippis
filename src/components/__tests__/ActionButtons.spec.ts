import {mount} from '@vue/test-utils';
import {describe, it, expect, beforeEach} from 'vitest';
import ActionButtons from '@/components/ActionButtons.vue';
import {useCocktailStore} from '@/stores/cocktail';
import {testCocktails, testSearchResults} from '@/components/__tests__/mswHandlers';

describe('ActionButtons', () => {
  const cocktailStore = useCocktailStore();

  beforeEach(() => {
    cocktailStore.$reset();
  });

  it('mounts', () => {
    const wrapper = mount(ActionButtons, {
      props: {
        cocktail: testSearchResults[0][1],
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('shows remove button', () => {
    cocktailStore.selection.push(testCocktails[0][1]);
    const wrapper = mount(ActionButtons, {
      props: {
        cocktail: testSearchResults[0][1],
      },
    });
    expect(wrapper.find('.remove-cocktail-0').exists()).toBe(true);
  });

  it('add button works', async () => {
    const wrapper = mount(ActionButtons, {
      props: {
        cocktail: testSearchResults[1][1],
      },
    });
    await wrapper.find('.add-cocktail-1').trigger('click');
    expect(cocktailStore.addToSelection).toHaveBeenCalledWith(1);
  });

  it('remove button works', async () => {
    cocktailStore.selection.push(testCocktails[0][1]);
    const wrapper = mount(ActionButtons, {
      props: {
        cocktail: testSearchResults[0][1],
      },
    });
    await wrapper.find('.remove-cocktail-0').trigger('click');
    expect(cocktailStore.removeFromSelection).toHaveBeenCalledWith(0);
  });
});
