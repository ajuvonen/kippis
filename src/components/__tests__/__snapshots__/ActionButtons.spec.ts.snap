// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ActionButtons > mounts 1`] = `
"<button data-v-634d56d3="" aria-label="Add &quot;Margarita&quot; to Selection" class="add-cocktail-0 action-button"><svg data-v-634d56d3="" viewBox="0 0 24 24" aria-hidden="true" class="w-5 h-5">
    <path d="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z" class="transition-all"></path>
  </svg></button>"
`;
